mm2src/coins/z_coin/z_rpc.rs

```diff
@@ -292,34 +350,111 @@ impl SaplingSyncLoopHandle {
     }
 
     async fn update_blocks_cache(&mut self) -> Result<(), MmError<UpdateBlocksCacheErr>> {
-        let request = tonic::Request::new(ChainSpec {});
-        let current_blockchain_block = self.grpc_client.get_latest_block(request).await?;
+        let current_block = match &mut self.rpc_client {
+            ZRpcClient::Native(client) => client.get_block_count().compat().await?,
+            ZRpcClient::Light(client) => {
+                let request = tonic::Request::new(ChainSpec {});
+                let block = client.get_latest_block(request).await?;
+                let res: u64 = block.into_inner().height;
+                res
+            },
+        };
         let current_block_in_db = block_in_place(|| self.blocks_db.get_latest_block())?;
-
         let from_block = current_block_in_db as u64 + 1;
-        let current_block: u64 = current_blockchain_block.into_inner().height;
-
         if current_block >= from_block {
-            let request = tonic::Request::new(BlockRange {
-                start: Some(BlockId {
-                    height: from_block,
-                    hash: Vec::new(),
-                }),
-                end: Some(BlockId {
-                    height: current_block,
-                    hash: Vec::new(),
-                }),
-            });
-
-            let mut response = self.grpc_client.get_block_range(request).await?;
-
-            while let Some(block) = response.get_mut().message().await? {
-                debug!("Got block {:?}", block);
-                block_in_place(|| self.blocks_db.insert_block(block.height as u32, block.encode_to_vec()))?;
-                self.notify_blocks_cache_status(block.height, current_block);
+            if let ZRpcClient::Light(client) = &mut self.rpc_client {
+                let request = tonic::Request::new(BlockRange {
+                    start: Some(BlockId {
+                        height: from_block,
+                        hash: Vec::new(),
+                    }),
+                    end: Some(BlockId {
+                        height: current_block,
+                        hash: Vec::new(),
+                    }),
+                });
+                let mut response = client.get_block_range(request).await?;
+                while let Some(block) = response.get_mut().message().await? {
+                    debug!("Got block {:?}", block);
+                    block_in_place(|| self.blocks_db.insert_block(block.height as u32, block.encode_to_vec()))?;
+                    self.notify_blocks_cache_status(block.height, current_block);
+                }
+            }
+            if let ZRpcClient::Native(client) = &self.rpc_client {
```

Comment:
I remember that we discussed this in person already, but still curious if there is a way to replace `if let` with `match` and also add a trait for ZRpc operations?

Reply:
Firstly about match pattern.
Light client requires `client` variable has mutable reference, because of method `get_block_range(request)`.
On the other hand working with `native client variable` we have method `get_block_by_height(height)`, which requires immutable reference of `client`. For instance, if I use match pattern with immutable reference, I will have to make light client variable &mut, which is [UB](https://stackoverflow.com/questions/54237610/is-there-a-way-to-make-an-immutable-reference-mutable) and another [link](https://doc.rust-lang.org/nomicon/transmutes.html).

Next trait implementation.
I was successful implementing the count of last block using trait `Client` its about methods `get_block_count()` of light client and `get_block_by_height()` of native. However, it is confusing on the step of getting block range, because we have to use `self.notify_blocks_cache_status()` (where self is self of structure `SaplingSyncLoopHandle`) in [while cycle ](https://github.com/KomodoPlatform/atomicDEX-API/blob/zcoin_native_mode_support/mm2src/coins/z_coin/z_rpc.rs#L365) and in [for cycle](https://github.com/KomodoPlatform/atomicDEX-API/blob/zcoin_native_mode_support/mm2src/coins/z_coin/z_rpc.rs#L365).
So it is extremely non-trivial to implement trait function for structure and use self of another structure inside the cycle. In my opinion, we wont get a gain in the amount of code by implementing the trait Client for light and native mode, because we have to do something with the function implementation of `SaplingSyncLoopHandle`.

Reply:
With my comment, I expected you to make additional research and refactoring attempt. Even if something seems "not-trivial", it's worth applying some effort to figure it out. We have to solve problems that are much more "not-trivial" than this. 

So here is the way to do it: https://github.com/KomodoPlatform/atomicDEX-API/commit/53c6bffaf2120bcb135c37423fd9ed6ba943f7d8 with some notes:
1. To avoid fighting with mut/immut borrowing, it's worth detaching the rpc_client from SaplingSyncLoopHandle making it a separate variable in the loop. 
2. I had to wrap `BlockDb` in `Mutex` to make it `Sync` and pass the `FnMut` closure to async fn from trait.

> we wont get a gain in the amount of code by implementing the trait Client for light and native 

In the mentioned commit, the specific implementations are cleanly separated. `update_blocks_cache` was over 100 lines, and with these `if let` statements, it could be already hard to read and understand. Also imagine if one day we will require a new variant to be added to `ZRpcClient` - with current approach, it would bloat `update_blocks_cache` even more. And with `if let`, it would be easy to miss the necessity to update `update_blocks_cache`, because there would be no compilation error without `match`.

The commit is not final, there's still some work required to fully remove `rpc_client` from `SaplingSyncLoopHandle`, and add other methods to trait. Also, there might be a way to make the code better, maybe avoid requiring Mutex on BlockDb. It's up to you to do further research on this :slightly_smiling_face:

Reply:
Thanks for the fair comments!
Trait implementation was added along with the necessary methods. So rpc_client was removed from `SaplingSyncLoopHandle`.
About database and `Mutex` - I need to think about it.